templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml

```diff
@@ -20,89 +20,116 @@ contents:
+            # Ensure resolv.conf exists and contains nameservers before we try to run podman
+            if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
+                cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
+            fi
 
-        NAMESERVER_IP=$(timeout 20s /usr/bin/podman run --rm \
-            --authfile /var/lib/kubelet/config.json \
-            --net=host \
-            {{ .Images.baremetalRuntimeCfgImage }} \
-            node-ip \
-            show \
-            --retry-on-failure \
-            {{range onPremPlatformAPIServerInternalIPs . }}"{{.}}" {{end}} \
-            {{range onPremPlatformIngressIPs . }}"{{.}}" {{end}})
-        DOMAINS="${IP4_DOMAINS} ${IP6_DOMAINS} {{.DNS.Spec.BaseDomain}}"
-        if [[ -n "$NAMESERVER_IP" ]]; then
-            if systemctl -q is-enabled systemd-resolved; then
-                >&2 echo "NM resolv-prepender: Setting up systemd-resolved for OKD domain and local IP"
-                if [[ ! -f /etc/systemd/resolved.conf.d/60-kni.conf ]]; then
+            # Pull container image outside of the timeout in a background process to workaround possible
+            # image corruption
+            # https://github.com/containers/podman/issues/14003
+            if ! /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; then
+                # Multiple podman pulls may occur in parallel wasting bandwidth therefore this
+                # needs to be removed when podman image corruption issue is resolved.
+                # Send to background to avoid SIGTERM from timeout which may cause image corruption issue during image pull
+                pull_baremetal_runtime_cfg_image &
+                # Wait until image is available
+                >&2 echo "NM resolv-prepender: Waiting for baremetal runtime cfg image to be available"
+                until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
+                    sleep 1
+                done
+                >&2 echo "NM resolv-prepender: Baremetal runtime cfg image is available"
+            fi
+
+            # Image must exist before reaching this point otherwise timeout may cause image corruption
+            NAMESERVER_IP=$(/usr/bin/podman run --rm \
+                --authfile /var/lib/kubelet/config.json \
+                --net=host \
+                {{ .Images.baremetalRuntimeCfgImage }} \
+                node-ip \
+                show \
+                --retry-on-failure \
+                {{range onPremPlatformAPIServerInternalIPs . }}"{{.}}" {{end}} \
+                {{range onPremPlatformIngressIPs . }}"{{.}}" {{end}})
+            DOMAINS="${IP4_DOMAINS} ${IP6_DOMAINS} {{.DNS.Spec.BaseDomain}}"
+            if [[ -n "$NAMESERVER_IP" ]]; then
+                if systemctl -q is-enabled systemd-resolved; then
+                    >&2 echo "NM resolv-prepender: Setting up systemd-resolved for OKD domain and local IP"
```

Comment:
FWIW, this was added by someone else (I think for OKD). I don't know that we test it at all in ci so historically I've tried to leave it alone.

I guess the only concern with this new logic is that systemd-resolved will be restarted every time the script runs. I don't know if that causes any disruption in DNS resolution, but if it does we might want to find some way to avoid that. Maybe we hash the file contents or something so we can check if they changed?

Reply:
We could write 60-kni.conf only if the contents are different, and compare the file tiemstamp with the service start timestamp.

But it does strike me as weird that the service needs to be restarted, there might be another option to make the service read the config without service interruption.

Another option is to set a null TERM trap handler for the portions of the script we don't want to be interrupted.

Reply:
I have not found a way to force systemd-resolved to read the configuration without a restart, but it does read it on its own a few seconds after the change. The only indication I found on the docs about this is 
```
SIGUSR2
Upon reception of the SIGUSR2 process signal systemd-resolved will flush all caches it maintains. Note that it should normally not be necessary to request this explicitly – except for debugging purposes – as systemd-resolved flushes the caches automatically anyway any time the host's network configuration changes. Sending this signal to systemd-resolved is equivalent to the resolvectl flush-caches command, however the latter is recommended since it operates in a synchronous way.
```

So perhaps we should not worry if we are killed between writing the configuration and restarting systemd, and just keep restarting systemd-resolved only if we actually wrote the configuration.